workout log to db progress

shriyan44 opened the pull request.

shriyan44 committed:
workout log to db progress

rohithdara commented on components/Home/Home.jsx:
```diff
@@ -101,8 +101,8 @@ export default () => {
       // Get the current logged in user id
-      const currentUser = firebase.auth().currentUser.uid;
-      // const currentUser = '68w6wWz8l5QJO3tDukh1fRXWYjD2';
+      // const currentUser = firebase.auth().currentUser.uid;
```
should switch back which line is commented before merging to main

rohithdara commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -77,6 +80,59 @@ const LogWorkout = (props) => {
+  const exerciseType = exercise.sets[0].reps ? 'REPS' : 'SECS';
+  return {
+    ID: exercise.exerciseId,
+    name: 'AYOOOOOO',
```
Is this temporary or why is the name AYOOOOOO? I'm confused lol

rohithdara reviewed (dismissed):
I tried hopping on your branch and logging the push workout. I remember you saying that the "undefined is not an object (evaluating 'set.time.toString')" was only broken on the arms one, but I got the same error for push since bench press set to time. Have you been able to find a fix? 
Other than that, everything else looks good!

shriyan44 committed:
Update components/LogWorkout/LogWorkout.jsx

jpoist97 commented on components/CreateCycle/CreateCycle.test.js:
```diff
@@ -0,0 +1,9 @@
+    const tree = renderer
+    .create(<EllipsisPopup></EllipsisPopup>).toJSON();
+    expect(tree).toMatchSnapshot();
+});
```
Did you mean to include this in the pr?

shriyan44 replied:
teehee

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -27,6 +30,7 @@ const parseExercises = (exercises) => exercises.map((exercise) => {
   const exerciseType = exercise.sets[0].reps ? 'REPS' : 'SECS';
 
   return {
+    ID: exercise.exerciseId,
```
```suggestion
    id: exercise.exerciseId,
```

Can we make this consistent with other code.

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -77,6 +81,56 @@ const LogWorkout = (props) => {
+      })
+  
+    }
+    userRef.collection('workouts').doc(workoutId).update(newWorkoutLog);
```
I don't think this currently updates lastPerformed, could you make sure that gets updated as well

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -167,8 +223,13 @@ LogWorkout.propTypes = {
   }).isRequired,
 };
 
+LogWorkout.defaultProps = {
+  isSelectedCycle: false,
+};
```
You need to put the route, params stuff outside for this to work as expected

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -77,6 +81,56 @@ const LogWorkout = (props) => {
+    <CurrentCycle
+          rightPress={() => { dispatch({ type: INCREMENT_SELECTED_CYCLE_INDEX, cycleLength: cycleDetails.length }); }}
+        /> )
+  };
```
I don't think this does what you want it to do. You just need a function that calls dispatch to increment selected cycle index and also updates the database as well. The stuff ou have commented out on line 91 is pretty close for the database part

jpoist97 reviewed (commented):
See comments, close to being done

shriyan44 committed:
selectedCycle workout updates properly

shriyan44 committed:
Merge branch 'main' of https://github.com/jpoist97/GainChanger into logworkout-db

shriyan44 committed:
progress on redux for prev details

shriyan44 committed:
Merge branch 'main' of https://github.com/jpoist97/GainChanger into logworkout-db

shriyan44 committed:
Merge branch 'main' of https://github.com/jpoist97/GainChanger into logworkout-db

shriyan44 committed:
fix conflict

shriyan44 committed:
attempting to fix cycle undefined error

shriyan44 committed:
everythang works

shriyan44 committed:
fixed lint

shriyan44 committed:
changed ID to id

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -76,6 +85,49 @@ const LogWorkout = (props) => {
+  const incrementSelectedCycleIdx = () => {
+    userRef.update({ selectedCycleIndex });
+    dispatch({ type: INCREMENT_SELECTED_CYCLE_INDEX, cycleLength });
+  };
```
What's the userRef.update doing here? To me this looks like it's not updating the selectedCycleIndex properly because you're sending it to firebase before incrementing it. Also if you want to access selectedCycleIndex you don't need to pass it through props, you can just access it directly from redux

jpoist97 replied:
Yeah I tested this and the selectedCycleIndex is not incremented in firebase, it'll work if you access it from redux put line 96 below the dispatch

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -76,6 +85,49 @@ const LogWorkout = (props) => {
+            weight: set.weight || set.prevWeight,
+          };
+          _.set(parsedSet, exercise.type === 'REPS' ? ['reps'] : ['time'], set.duration || set.prevDuration);
+          return parsedSet;
```
nice

jpoist97 commented on reducers/workoutReducer.js:
```diff
@@ -14,6 +14,9 @@ const workoutReducer = (state = initialState, action) => {
          console.log(`Adding workout to store with ${action.workout}`);
          const newWorkouts = [...state.workouts];
+         newWorkouts.map((workoutEle) => {
+            console.log(workoutEle.id)
```
Delete this

jpoist97 commented on reducers/workoutReducer.js:
```diff
@@ -24,6 +27,24 @@ const workoutReducer = (state = initialState, action) => {
+         });
+         return {
+            workouts: newW
+         }
```
Can we rename newW to something better

jpoist97 commented on actions/workoutActions.js:
```diff
@@ -21,8 +21,16 @@ const deleteWorkout = (workoutId) => {
+      type: UPDATE_WORKOUT_PREV,
+      workout,
+   }
+}
```
Can we change this to just take in **workoutId and updatedExercises**. Those are the only two fields you use from the workout and it'll be confusing later since we'll think this requires an entire workout object

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -76,6 +85,49 @@ const LogWorkout = (props) => {
+      exercises: newWorkoutLog.exercises,
+    };
+
+    userRef.collection('workouts').doc(workoutId).update(newWorkoutDoc); // updates workout doc
```
This code stores extra information in firebase, we don't want the workoutId to be stored here

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -76,6 +85,49 @@ const LogWorkout = (props) => {
+    const workoutRecsRef = userRef.collection('workoutRecords');
+
+    const newWorkoutLog = {
+      workoutName: name,
```
Workout logs do not need name

jpoist97 replied:
Just kidding actually, I think we should store the workout name and exercise name in the workout log so Rohith and Fin don't need to cross-reference them in the calendar component

jpoist97 reviewed (commented):
Looking good, see comments for things to fix

jpoist97 commented:
Also general point, did we say we didn't want people to be able to log a workout unless everything was completed

shriyan44 committed:
trying to unbreak branch

shriyan44 committed:
everything works again!

still need to implement part to only be able to log when all sets r completed tho

shriyan44 committed:
rohith date change

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -99,12 +99,23 @@ const LogWorkout = (props) => {
+        }
+      }
+      return date.getFullYear() + '-' + formatMonth(date.getMonth() + 1) + '-' + date.getDate();
+    }
```
Use date-fns to format the date rather than writing it yourself

rohithdara replied:
lol I wrote that, but I sent a date-fns version to replace this with

jpoist97 reviewed (commented):
Code looks good, I think you're just missing the alert when someone tries to log workout and they haven't completed everything yet

shriyan44 committed:
rohith date change 2

shriyan44 committed:
only update db if sets completed

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -76,6 +86,49 @@ const LogWorkout = (props) => {
+
+  const sendWorkoutLogToDB = () => {
+    const workoutRecsRef = userRef.collection('workoutRecords');
+    const { format } = require('date-fns');
```
Can you change this to an import at the top of the code

shriyan44 replied:
sorry dont understand what u want

jpoist97 replied:
instead of `const { format } = require('date-fns')` here in the middle of your code move it to the top and import it like all of the other modules used in this file. `import { format } from 'date-fns'`

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -146,7 +199,19 @@ const LogWorkout = (props) => {
+            completedList.push(set.completed);
+          });
+        });
+        if (completedList.includes(false)) {
```
I don't think this is the best way to do this. I think we could replace this with nested for loops that exit on the first false they find. That way if the first one is false we don't waste time on the rest of them and we don't need to call another .includes after to check

jpoist97 replied:
Also can you take out the TODO now that you've implemented that functionality

shriyan44 replied:
I get what u mean and that was kinda my first way of doing it so I had like an if false then show alert and return, but I was confused on the logic cuz I didnt know how to do it so that the finish functions only get called if all are completed

jpoist97 replied:
There's a number of ways you can do it. Here's one off the top of my head

```
let completed = true;
for(let i = 0; completed && i < exerciseState.length; i++) {
   const sets = exerciseState[i].exercise.sets;
   for(let j = 0; completed && j < sets.length; j++) {
      if(!set[j].completed) {
         completed = false;
      }
   }
}

if(!completed) {
   <Show alert>
} else {
   <Do database stuff>
}

```

It might not work exactly correctly but the logic is there

shriyan44 replied:
wow I rlly forgot about flags that was so incredibly simple my b lmao

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -76,6 +86,49 @@ const LogWorkout = (props) => {
+
+  const incrementSelectedCycleIdx = () => {
+    dispatch({ type: INCREMENT_SELECTED_CYCLE_INDEX, cycleLength });
+    userRef.update({ selectedCycleIndex: (cycleIdx + 1) % cycleLength });
```
Can you flip these two lines. I know it works this way but I'm a bit worried that since you're using the actual selectedCycleId this might increment it 2 times in the database (even though it doesn't). It's just weird to me that it's not being re-rendered after your action completes.

jpoist97 reviewed (commented):
Look good! I'm excited to use the app once this goes in. One other thing is did Fin and Rohith want you to also add to like a list of dates in the user when they log a workout?

rohithdara commented:
Fin and I have that list working in our PR, so no need for Shriya to do it

jpoist97 commented:
Also this shouldn't post any data to firestore as "n/a", it should be null if the value is "n/a"

shriyan44 committed:
send fb null instead of n/a

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -1,7 +1,7 @@
 import PropTypes from 'prop-types';
-import _ from 'lodash';
+import _, { set } from 'lodash';
 import styled from 'styled-components';
```
Remove { set } and test this again. I think this is gonna be used instead of the actual set inside your map

jpoist97 commented on components/LogWorkout/LogWorkout.jsx:
```diff
@@ -111,9 +112,9 @@ const LogWorkout = (props) => {
         sets: exercise.sets.map((set) => {
           const parsedSet = {
-            weight: set.weight || set.prevWeight,
+            weight: (set.prevWeight == 'n/a' && set.weight  == '') ? null : parseInt(set.weight) || parseInt(set.prevWeight),
```
Use triple equals in javascript please.

Would something like this work better:
`parseInt(set.weight) || set.prevWeight === 'n/a' ? null : parseInt(set.prevWeight)`

This way we only check that prevWeight is n/a when we have to. Also I'm p sure NaN is falsey so you don't need to worry about set.weight being an empty string.

shriyan44 replied:
actually I tried your way and it doesnt work it was sending null when I set a weight so I just changed it back to mine

jpoist97 reviewed (commented):
Also please address my old comments and fix the build

shriyan44 commented:
> Look good! I'm excited to use the app once this goes in. One other thing is did Fin and Rohith want you to also add to like a list of dates in the user when they log a workout?

  just modified the format that the date was being sent in idk bout no list

jpoist97 commented:
> > Look good! I'm excited to use the app once this goes in. One other thing is did Fin and Rohith want you to also add to like a list of dates in the user when they log a workout?
> 
> just modified the format that the date was being sent in idk bout no list

Rohith said they did that in their pr so no need for you to do it

shriyan44 committed:
am i done yet :(

jpoist97 reviewed (dismissed):
Done, just need to fix merge conflicts

shriyan44 committed:
Merge branch 'main' of https://github.com/jpoist97/GainChanger into logworkout-db

merge rohith and fins code
